### PR TITLE
#3282 - Detect Assessment Offering Dates Change (Part 1): DB Migrations

### PR DESCRIPTION
**Pull request by sh16011993**
## As a part of this PR, the following was completed:

- Created new columns on the `sims.student_assessments` table: `previous_date_changed_reported_assessment_id` and `reported_date` columns.
- Added the columns to the `student-assessment.model.ts` file.

### CODE REVIEW

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-reported-date-col.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE
+  sims.student_assessments
+ADD
+  COLUMN reported_date TIMESTAMPTZ;
+
+COMMENT ON COLUMN sims.student_assessments.reported_date IS 'Timestamp for the last reported assessment id.';
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-reported-date-col.sql:6]: Please avoid technical terms like timestamp in comment. This column indicates the date at which the assessment causing the study date change has been reported.

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+ALTER table
+  sims.student_assessments
+ADD
+  COLUMN previous_date_changed_reported_assessment_id INT REFERENCES sims.student_assessments(id);
+
+COMMENT ON COLUMN sims.student_assessments.previous_date_changed_reported_assessment_id IS 'Last reported assessment id corresponding to the offering date change.';
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql:6]: What about `Assessment that changed the study dates of the application after previously reported assessment.` ?

Comment (dheepak-aot) [inline sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql:6]: Not a blocker. Is there any benefit of using separate .sql files for adding the 2 columns ?

Comment (sh16011993) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql:6]: I can put them in the same sql file.

**Comment by dheepak-aot** [comment]
Nice work. Added few minor comments @sh16011993

File: sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts [hunk 3]
```diff
@@ -260,6 +260,23 @@ export class StudentAssessment extends RecordDataModel {
...
+  /**
+   * Last reported assessment id corresponding to the offering date change.
+   */
+  @Column({
+    name: "previous_date_changed_reported_assessment_id",
+    nullable: true,
+  })
+  previousDateChangedReportedAssessmentId?: number;
+  /**
+   * Timestamp for the last reported assessment id.
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts:272]: Model?

File: sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/db-migrations/src/sql/StudentAssessments/Add-previous-date-changed-reported-assessment-id-col.sql]: Thanks for updating the description further as my suggested one was not accurate.

File: sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts [hunk 4]
```diff
@@ -260,6 +260,24 @@ export class StudentAssessment extends RecordDataModel {
     referencedColumnName: ColumnNames.ID,
   })
   relatedApplicationAssessment?: StudentAssessment;
+  /**
+   * Last reported assessment id corresponding to the offering date change.
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts:264]: can you remove the id the the comment?

File: sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts [hunk 5]
```diff
@@ -260,6 +260,24 @@ export class StudentAssessment extends RecordDataModel {
...
+   */
+  @ManyToOne(() => StudentAssessment, { nullable: true })
+  @JoinColumn({
+    name: "previous_date_changed_reported_assessment_id",
+    referencedColumnName: ColumnNames.ID,
+  })
+  previousDateChangedReportedAssessment?: StudentAssessment;
+  /**
+   * Timestamp for the last reported assessment id.
+   */
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts:274]: Same here. Please update the comment matching the SQL column comment.?

File: sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts [hunk 6]
```diff
@@ -260,6 +260,24 @@ export class StudentAssessment extends RecordDataModel {
     referencedColumnName: ColumnNames.ID,
   })
   relatedApplicationAssessment?: StudentAssessment;
+  /**
+   * Last reported assessment id corresponding to the offering date change.
+   */
+  @ManyToOne(() => StudentAssessment, { nullable: true })
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/student-assessment.model.ts:266]: I notice that eager and cascade are not set to false(we are setting it explicitly in most  instances in our code) and from the documentation I do not see any side effect of not setting it as false. 

@andrewsignori-aot @guru-aot @andrepestana-aot @lewischen-aot FYI.

**Review by dheepak-aot — APPROVED** [review]
Thanks for making the changes @sh16011993. 👍

**Pull request merged**